idrstream/merge_CP_DP.py

```diff
@@ -0,0 +1,162 @@
+"""
+This file holds the functions necessary for merging IDR_stream CP and DP outputs.
+As the IDR_stream output follows the pycytominer output format, these merge functions should work for any pycytominer output data.
+"""
+
+import pandas as pd
+import pathlib
+import math
+
+
+def merge_CP_DP_batch_data(
+    cp_batch_data: pd.DataFrame, dp_batch_data: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    merge dataframes for IDR_stream output with CP and DP features
+    the two features dataframes should have aligned location metadata (plate, well, frame, etc) and the same number of rows (cells)
+
+    Parameters
+    ----------
+    cp_batch_data : pd.DataFrame
+        idrstream_cp batch output
+    dp_batch_data : pd.DataFrame
+        idrstream_dp batch output
+
+    Returns
+    -------
+    pd.DataFrame
+        merged batch data with metadata, CP features, and DP features
+
+    Raises
+    ------
+    IndexError
+        cp and dp dataframes have different number of rows (cells)
+    """
+
+    # check batch data have same number of rows (cells)
+    # if batch data have different number of cells, raise an error because they must not have close segmentations
+    if cp_batch_data.shape[0] != dp_batch_data.shape[0]:
+        raise IndexError("Batch data have different number of rows (cells)!")
+
+    # hide warning for pandas chained assignment
+    pd.options.mode.chained_assignment = None
+
+    # get cp and dp column names
+    cp_columns = cp_batch_data.columns
+    dp_columns = dp_batch_data.columns
+    # get metadata columns (columns that show up in both dataframes)
+    metadata_columns = [col for col in cp_columns if col in dp_columns]
+
+    # remove metadata columns from cp and dp columns
+    cp_columns = set(cp_columns) - set(metadata_columns)
+    dp_columns = set(dp_columns) - set(metadata_columns)
+
+    # add CP and DP prefixes to their respective columns
+    cp_batch_data = cp_batch_data.rename(
+        columns={col: f"CP__{col}" for col in cp_columns}
+    )
+    dp_batch_data = dp_batch_data.rename(
+        columns={col: f"DP__{col}" for col in dp_columns}
+    )
+
+    # get each image path because cells within the same image are trying to be associated
+    image_paths = cp_batch_data["Metadata_DNA"].unique()
+
+    # list to compile merged dataframes from each image
+    compiled_merged_data = []
+
+    # iterate through each image to focus on matching cell positions (x,y)
+    for image_path in image_paths:
```

Comment:
Could these merges be performed without this loop, making references where appropriate to each dataset's unique `image_path` values?

Reply:
Although it would be possible to complete this merge without the loop, I think the for loop makes the code very readable and understandable and is worth keeping for the merge function.